File: 98_ValidateBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        current = root
        stack = []
        visited = []
        while True:
            if current is not None:
                stack.append(current)
                current = current.left
            elif(stack):
                current = stack.pop()
                if visited:
                    if visited[-1] >= current.val:
                        return False
                visited.append(current.val)
                current = current.right
            else:
                break
        return True

File: C++/98_ValidateBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        TreeNode* current = root;
        stack<TreeNode*> stack;
        vector<int> visited;
        while(true){
            if (current != NULL){
                stack.push(current);
                current = current->left;
            } else if (!stack.empty()){
                current = stack.top();
                stack.pop();
                if (visited.size() != 0){
                    if (visited[visited.size()-1] >= current->val){
                        return false;
                    }
                }
                visited.push_back(current->val);
                current = current->right;
            } else{
                break;
            }
        }
        return true;
    }
};